homework10/task1/main.c

```diff
@@ -0,0 +1,171 @@
+            ++i;
+            if (i >= cities) {
+                i = 1;
+            }
+        }
+        conquer(capitalsCities[i], testMatrix, testMarked, cities);
+        ++i;
+        if (i >= cities) {
+            i = 1;
+        }
+    }
+    free(testMarked);
+    free(testMatrix);
+    return listLength(capitalsCities[1]) == 2 && listLength(capitalsCities[3]) == 3;
+}
+
+int main(void) {
+    if (!testMatrix()) {
+        printf("Test failed!\n");
+        return -1;
+    }
+    FILE* file = fopen("file.txt", "r");
+    int cities = 0;
+    fscanf(file, "%d", &cities);
+    int roads = 0;
+    fscanf(file, "%d", &roads);
+    int** matrix = calloc(cities + 1, sizeof(int*));
+    for (int i = 1; i <= cities; ++i) {
+        matrix[i] = calloc(cities, sizeof(int*));
+    }
+    for (int i = 0; i < roads; ++i) {
+        int row = 0;
+        int column = 0;
+        int length = 0;
+        fscanf(file, "%d", &row);
+        fscanf(file, "%d", &column);
+        fscanf(file, "%d", &length);
+        matrix[row][column] = matrix[column][row] = length;
+    }
+    int capitalsAmount = 0;
+    fscanf(file, "%d", &capitalsAmount);
+    List** citiesOfCountry = calloc(cities + 1, sizeof(List*));
+    if (citiesOfCountry == NULL) {
+        return -1;
+    }
+    bool* marked = calloc(cities + 1, sizeof(bool));
+    for (int i = 0; i < capitalsAmount; i++) {
+        int capital = 0;
+        fscanf(file, "%d", &capital);
+        citiesOfCountry[capital] = createList();
+        insert(capital, citiesOfCountry[capital]);
+        marked[capital] = true;
+    }
+    int i = 0;
+    while (!allMarked(marked, cities + 1)) {//we check if we have visited all the cities
+        if (i >= cities) {
+            i = 1;
+        }
+        while (citiesOfCountry[i + 1] == NULL) {
+            ++i;
+            if (i >= cities) {
+                i = 0;
+            }
+        }
+        conquer(citiesOfCountry[i + 1], matrix, marked, cities);
+        ++i;
+        if (i >= cities) {
+            i = 0;
+        }
+    }
+    printMatrix(matrix, cities);
+    for (int i = 1, j = 1; i < cities + 1; ++i) {
+        if (citiesOfCountry[i] != NULL) {
+            printf("Cities of %d country: ", j);
+            printList(citiesOfCountry[i]);
+            ++j;
+        }
+    }
+    free(marked);
+    free(matrix);
+    fclose(file);
+    return 0;
+}
+
+//outputs the matrix to the console
+void printMatrix(int** matrix, int cities) {
+    printf(" ");
+    for (int i = 1; i < cities + 1; ++i) {
+        printf(" %d", i);
+    }
+    printf("\n");
+    for (int i = 1; i < cities + 1; ++i) {
+        printf("%d|", i);
+        for (int j = 1; j < cities + 1; ++j) {
+            printf("%d ", matrix[i][j]);
+        }
+        printf("\n");
+    }
+}
+
+//checks if we have visited all the cities
+bool allMarked(bool marked[], int length) {
+    for (int i = 1; i < length; ++i) {
+        if (marked[i] == false) {
```

Comment:
```suggestion
        if (!marked[i]) {
```
Стайлгайд запрещает явные сравнения с булевыми константами